Update heatmap and toolbar docs (#6093)
PostHog/posthog.com: contents/docs/product-analytics/toolbar.mdx

    enterprise: full
---

The toolbar is like "Inspect Element" for PostHog features and data. When enabled, the toolbar appears as an overlay **on your product or website** (it's not visible to your users). With it, you can interact with elements to [create actions](/tutorials/how-to-capture-events-the-easy-ways), visualize heatmaps, override feature flags, and more.

![PostHog toolbar](../../images/docs/product-analytics/toolbar.png)

## How to launch the toolbar

> 1. In your PostHog instance, Go to '[Launch Toolbar](https://app.posthog.com/toolbar)' in the left-hand menu

Reviewer: ```suggestion
1. In your PostHog instance, go to '[Launch Toolbar](https://app.posthog.com/toolbar)' in the left-hand menu
```

2. Click 'Enable PostHog Toolbar'
3. Add the URLs you wish to enable the toolbar for. Only you will be able to see it – not your users.
4. Click Save and then Launch.    
5. A new window will open with toolbar on your website. Click on the toolbar to interact with it.

> ![How to setup the PostHog toolbar](../../images/docs/product-analytics/toolbar-setup.mp4)

Reviewer 2: This isn't loading for me for some reason.

Author: Not sure why, but I checked now and its showing in the Vercel preview - https://posthog-git-update-heatmap-and-toolbar-docs-post-hog.vercel.app/docs/product-analytics/toolbar

## Heatmaps 

Heatmaps is a feature in the toolbar that shows you how users are interacting with elements on your website or app. 

> **Note:** To use heatmaps, [autocapture](/docs/data/autocapture) must be enabled.

To view your heatmaps, click on the toolbar and then click on the red fire (🔥) icon.

![Launch Toolbar](../../images/tutorials/toolbar/toolbar-options.png)

With the heatmap on, clickable elements on your website will have a red overlay with numbers on them. Here's an example of our website with the heatmap enabled:

![An example from the PostHog website showing the heatmap highlighting a number of page elements. Some showing a count of clicks and one also showing a rage-click](../../images/tutorials/toolbar/posthog-heatmap-example.png)

> Each number in the top right box of each element represents the number of clicks it received. The number in the top left of the element is the number of rage-clicks. 

Author: @ivanagas Not sure I follow why remove this?

> Each number in the top right box of each element represents the number of clicks it received. The number in the top left of the element is the number of rage-clicks. 

Reviewer 2: It didn't seen this as actually being the case, on our homepage we are missing it for example.

> Each number in the top right box of each element represents the number of clicks it received. The number in the top left of the element is the number of rage-clicks. 

Reviewer 3: ![2023-06-12 20 36 06](https://github.com/PostHog/posthog.com/assets/984817/db457630-431f-4894-b82a-3bdfd8487ecc)

The first example I found is terrible... the little drop down button has rageclicks

Numbers at the bottom or top of the screen indicate that there are more elements that are not currently being displayed on the page.

### How to create an action from a heatmap

Once you have enabled the heatmap, click on any element. You will then be able to create an [action](/docs/data/actions) from it.

> ![How to create an action using heatmaps](../../images/products/product-analytics/heatmaps-create-action.mp4)

Reviewer 3: each of the mp4 links don't show in GitHub's markdown preview... is it just because GitHub doesn't know how to render them?

Reviewer 3: yep, they work in the vercel preview ✅

> ### How to create an action from a heatmap

Reviewer 2: Move this section above change date time, so you have all the heatmap menu related content together.

![The selector modal being used to discover the more meaningful selector](../../images/tutorials/toolbar/toolbar-selector-edit.gif)

## Viewing your existing actions

You can view your existing actions on your page by clicking the "Actions" button on the Toolbar. This will highlight any elements that are associated with an action. This is a useful way to check that elements are associated with the correct actions.

By clicking on a specific element, you can update or delete the action.

Note that the number in the green box on the top left of the element is the action's index in your list of actions (and does not represent how many times that action was triggered). The number in the yellow box next to the "Actions" button in the toolbar represents the total number of actions you have created.

![How to setup the PostHog toolbar](../../images/products/product-analytics/toolbar-view-actions.mp4)

## Overriding feature flags with the toolbar

The toolbar also enables you to test your feature flags. You can enable, disable, or override your feature flags, and then view how your website or app changes with the new feature flags values.

> To do this, click on the "Feature Flags" button in the toolbar. Then, search for any feature flag of choice and click on the toggles to change it's value. 

Reviewer 2: ```suggestion
To do this, click on the "Feature Flags" button in the toolbar, search for any feature flag of choice ,and click on the toggles to change its value. 
```

Note that overriding feature flags will only affect *your* browser. You may also need to refresh the page to see how your change affect your website.

![How to override feature flags using the tool](../../images/products/product-analytics/toolbar-feature-flags.mp4) 

## Troubleshooting and FAQ

### Toolbar not loading or displaying

When launched, the toolbar injects an API token in the format `#__posthog=JSON` in your site's URL, like this: `http://mysite.com/#__posthog=JSON`.

The toolbar won't load if you have a frontend that overrides the injected API token (the information after the #) before the `posthog-js` library (or snippet) loads and has a chance to read the API token.

> To solve this, retrieve the `__posthog` JSON from the URL before it is overridden and call `posthog.loadToolbar(JSON)`:

Author: hmm good point. I actually tried to do this now and was not able to. @pauldambra can share more details on how to do this exactly?

I tried in the browser console, but ran into "posthog not defined" issue

> To solve this, retrieve the `__posthog` JSON from the URL before it is overridden and call `posthog.loadToolbar(JSON)`:

Reviewer 2: I think @mariusandra implemented this?

> To solve this, retrieve the `__posthog` JSON from the URL before it is overridden and call `posthog.loadToolbar(JSON)`:

Reviewer 4: You'd call this in your code, wherever you have access to this variable in your project's frontend router.

I think in our case, `posthog` is not available in the console because we use the npm package.
